design/DESIGN_typed_context_pipeline.md

```diff
@@ -0,0 +1,451 @@
+# Design: Typed Context Pipeline for Relic
+
+## 1. Introduction and Goals
+
+This document proposes a design for a typed context pipeline in Relic. The primary goal is to achieve **compile-time safety** for middleware and handler composition. This means the Dart analyzer should be able to verify that if a handler (or subsequent middleware) expects certain data to be present in its request context (e.g., an authenticated `User` object), the necessary preceding middleware (e.g., an `AuthenticationMiddleware`) has been correctly configured in the pipeline.
+
+This approach aims to:
+*   Catch pipeline configuration errors at compile-time rather than runtime.
+*   Improve developer ergonomics by providing clear, type-safe access to context data.
+*   Maintain high performance by leveraging zero-cost abstractions like extension types.
+*   Avoid a "class explosion" that might occur if every combination of context data required a distinct context class.
+
+## 2. Core Components
+
+The system relies on a few key components:
+
+### 2.1. Relic's `RequestContext` and Stable Request Token
+
+This design utilizes Relic's existing `RequestContext` (defined in `relic/lib/src/adapter/context.dart`). The key aspects of `RequestContext` relevant to this design are:
+*   It provides access to the current `Request` object (e.g., via a `request` getter).
+*   It contains a stable `token` (e.g., via a `token` getter). This `token` is a unique identifier for the entire lifecycle of a single request and remains constant even if the `RequestContext` instance itself undergoes state transitions (e.g., from `NewContext` to `ResponseContext`).
+
+All request-scoped data attached via `Expando` (managed by `ContextProperty`) will be keyed off this stable `requestToken`. The extension type views defined in this design will wrap an instance of Relic's `RequestContext`.
```

Comment:
Not sure about the performance characteristics with using the various expandos. I think we could similarly have just the state on the context / token itself, and then provide getter / setter like `V get(ContextProperty<V> prop)` on it.

I like how easy it is in the current approach to hide the setter though (just making that part file-local for example), and wonder if replicating this would be much harder in my above suggestion (e.g. you could have something like `Completer`, with an "value reader" and "value setter" being distinct, and only 1 exported as a direct helper).